File: advent-of-code/2025/day-7-laboratories/normal/n-d07-12.kt
fun main() {
    val grid = generateSequence(::readLine).toList()
    val h = grid.size
    val w = grid[0].length
    val sx = (0 until w).first { grid[0][it] == 'S' }
    var splits = 0
    val queue = ArrayDeque<Pair<Int, Int>>()
    val visited = mutableSetOf<Pair<Int, Int>>()
    queue.add(0 to sx)
    while (queue.isNotEmpty()) {
        val (y, x) = queue.removeFirst()
        if ((y to x) in visited) continue
        visited.add(y to x)
        val ny = y + 1
        if (ny >= h) continue
        if (grid[ny][x] == '^') {
            splits++
            if (x > 0) queue.add(ny to x - 1)
            if (x < w - 1) queue.add(ny to x + 1)
        } else queue.add(ny to x)
    }
    
    val memo = mutableMapOf<Pair<Int, Int>, Long>()
    fun countPaths(y: Int, x: Int): Long {
        if (y >= h) return 1
        memo[y to x]?.let { return it }
        val ny = y + 1
        if (ny >= h) return 1
        val result = if (grid[ny][x] == '^') {
            (if (x > 0) countPaths(ny, x - 1) else 0) + (if (x < w - 1) countPaths(ny, x + 1) else 0)
        } else countPaths(ny, x)
        memo[y to x] = result
        return result
    }
    
    println(splits)
    println(countPaths(0, sx))
}

File: advent-of-code/2025/day-7-laboratories/normal/n-d07-12.rs
use std::io::{self, BufRead};
use std::collections: :{HashMap, HashSet, VecDeque};

fn main() {
    let grid: Vec<Vec<char>> = io::stdin().lock().lines().flatten().map(|l| l.chars().collect()).collect();
    let (h, w) = (grid.len(), grid[0].len());
    let sx = (0.. w).find(|&x| grid[0][x] == 'S').unwrap();
    let mut splits = 0;
    let mut queue = VecDeque::new();
    let mut visited = HashSet::new();
    queue.push_back((0, sx));
    while let Some((y, x)) = queue.pop_front() {
        if visited.contains(&(y, x)) { continue; }
        visited.insert((y, x));
        let ny = y + 1;
        if ny >= h { continue; }
        if grid[ny][x] == '^' {
            splits += 1;
            if x > 0 { queue.push_back((ny, x - 1)); }
            if x < w - 1 { queue.push_back((ny, x + 1)); }
        } else { queue.push_back((ny, x)); }
    }
    fn count_paths(y: usize, x: usize, h: usize, w: usize, grid: &Vec<Vec<char>>, memo: &mut HashMap<(usize, usize), i64>) -> i64 {
        if y >= h { return 1; }
        if let Some(&v) = memo.get(&(y, x)) { return v; }
        let ny = y + 1;
        if ny >= h { return 1; }
        let result = if grid[ny][x] == '^' {
            (if x > 0 { count_paths(ny, x - 1, h, w, grid, memo) } else { 0 }) +
            (if x < w - 1 { count_paths(ny, x + 1, h, w, grid, memo) } else { 0 })
        } else { count_paths(ny, x, h, w, grid, memo) };
        memo.insert((y, x), result);
        result
    }
    
    let mut memo = HashMap::new();
    println!("{}\n{}", splits, count_paths(0, sx, h, w, &grid, &mut memo));
}
